libbpf-tools/xfsslower.bpf.c

```diff
@@ -0,0 +1,159 @@
+const volatile __u64 min_lat = 0;
+
+struct piddata {
+	u64 ts;
+	loff_t start;
+	loff_t end;
+	struct file *fp;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 8192);
+	__type(key, u32);
+	__type(value, struct piddata);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+static __always_inline int
+probe_entry(struct file *fp, loff_t s, loff_t e)
+{
+	u64 id = bpf_get_current_pid_tgid();
+	struct piddata piddata = {};
+	u32 tgid = id >> 32;
+	u32 pid = id;
+
+	if (targ_tgid && targ_tgid != tgid)
+		return 0;
+
+	piddata.ts = bpf_ktime_get_ns();
+	piddata.start = s;
+	piddata.end = e;
+	piddata.fp = fp;
+	if (piddata.fp)
+		bpf_map_update_elem(&start, &pid, &piddata, 0);
+	return 0;
+}
+
+SEC("kprobe/xfs_file_read_iter")
+int BPF_KPROBE(kprobe__xfs_file_read_iter, struct kiocb *iocb,
+	       struct iov_iter *to)
+{
+	struct file *fp = BPF_CORE_READ(iocb, ki_filp);
+	loff_t start = BPF_CORE_READ(iocb, ki_pos);
+
+	return 0;
+}
+
+SEC("kprobe/xfs_file_write_iter")
+int BPF_KPROBE(kprobe__xfs_file_write_iter, struct kiocb *iocb,
+	       struct iov_iter *from)
+{
+	struct file *fp = BPF_CORE_READ(iocb, ki_filp);
+	loff_t start = BPF_CORE_READ(iocb, ki_pos);
+
+	return probe_entry(fp, start, 0);
+}
+
+SEC("kprobe/xfs_file_open")
+int BPF_KPROBE(kprobe__xfs_file_open, struct inode *inode, struct file *file)
+{
+	return probe_entry(file, 0, 0);
+}
+
+SEC("kprobe/xfs_file_fsync")
+int BPF_KPROBE(kprobe__xfs_file_fsync, struct file *file, loff_t start,
+	       loff_t end, int fdatasync)
+{
+	return probe_entry(file, start, end);
+}
+
+static __always_inline int probe_exit(struct pt_regs *ctx, char type)
+{
+	u64 id = bpf_get_current_pid_tgid();
+	u64 end_ns = bpf_ktime_get_ns();
+	struct piddata *piddatap;
+	struct event event = {};
+	struct file *fp = NULL;
+	struct qstr qs = {};
+	u32 tgid = id >> 32;
+	u32 pid = id;
+	u64 delta_us;
+
+	if (targ_tgid && targ_tgid != tgid)
+		return 0;
+
+	piddatap = bpf_map_lookup_elem(&start, &pid);
+	if (!piddatap)
+		return 0;    /* missed entry */
+
+	delta_us = (end_ns - piddatap->ts) / 1000;
+	bpf_map_delete_elem(&start, &pid);
+
+	if ((s64)delta_us < 0 || delta_us <= min_lat * 1000)
+		return 0;
+
+	fp = piddatap->fp;
+	qs = BPF_CORE_READ(fp, f_path.dentry, d_name);
+	if (qs.len == 0)
+		return 0;
+
+	bpf_core_read_str(&event.file, sizeof(event.file), qs.name);
+
+	bpf_get_current_comm(&event.task, sizeof(event.task));
+	event.delta_us = delta_us;
+	event.end_ns = end_ns;
+	event.offset = piddatap->start;
+	event.size = type != TRACE_FSYNC ? PT_REGS_RC(ctx) :
+		piddatap->end - piddatap->start;
+	event.type = type;
+	event.tgid = tgid;
+
+	/* output */
+	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU,
+			      &event, sizeof(event));
+	return 0;
+}
+
+SEC("kretprobe/xfs_file_read_iter")
+int BPF_KRETPROBE(kretprobe__xfs_file_read_iters)
```

Comment:
you can add `int ret` arg into BPF_KRETPROBE and pass it into probe_exit directly. That way you won't need explicit PT_REGS_RC() in prove_exit.

Reply:
@anakryiko I had a doubt about `int ret` from `BPF_KRETPROBE`'s define 

```c
#define ___bpf_kretprobe_args0() ctx
#define ___bpf_kretprobe_args1(x) \
	___bpf_kretprobe_args0(), (void *)PT_REGS_RET(ctx)
#define ___bpf_kretprobe_args(args...) \
	___bpf_apply(___bpf_kretprobe_args, ___bpf_narg(args))(args)

/*
 * BPF_KRETPROBE is similar to BPF_KPROBE, except, it only provides optional
 * return value (in addition to `struct pt_regs *ctx`), but no input
 * arguments, because they will be clobbered by the time probed function
 * returns.
 */
#define BPF_KRETPROBE(name, args...)					    \
name(struct pt_regs *ctx);						    \
static __attribute__((always_inline)) typeof(name(0))			    \
____##name(struct pt_regs *ctx, ##args);				    \
typeof(name(0)) name(struct pt_regs *ctx)				    \
{									    \
	_Pragma("GCC diagnostic push")					    \
	_Pragma("GCC diagnostic ignored \"-Wint-conversion\"")		    \
	return ____##name(___bpf_kretprobe_args(args));			    \
	_Pragma("GCC diagnostic pop")					    \
}									    \
static __always_inline typeof(name(0)) ____##name(struct pt_regs *ctx, ##args)
```
It looks like:
```c
int ret = ctx->sp;  // ___bpf_kretprobe_args0(), (void *)PT_REGS_RET(ctx)
```
not 
```c
int ret = ctx->ax;  // PT_REGS_RC(ctx)
```
https://github.com/libbpf/libbpf/blob/master/src/bpf_tracing.h#L58
https://github.com/libbpf/libbpf/blob/master/src/bpf_tracing.h#L60

Is this a bug?

Reply:
Yeah, clearly a bug, it should use PT_REGS_RC, not PT_REGS_RET, my bad. Feel free to submit a patch, if not, I'll do it tomorrow or Monday.

Reply:
Ok, I'll submit it.